Source/d_main.c

```diff
@@ -743,281 +747,99 @@ static void PrepareAutoloadPaths (void)
-    if (strncmp(n,"REKCREDS",8) == 0)
-      ++rekkr;
-}
+    int count = 0;
+    filelump_t lump;
+    wadinfo_t header;
 
-  fclose(fp);
-
-  *gmission = doom;
-  *hassec = false;
-  *gmode =
-    // [FG] improve gamemission detection to support the Freedoom IWADs
-    cm >= 30 ? (*gmission = (tnt >= 4 && plut < 8) ? pack_tnt :
-                (plut >= 8 && tnt < 4) ? pack_plut : doom2,
-                *hassec = sc >= 2, commercial) :
-    ud >= 9 ? retail :
-    rg >= 18 ? registered :
-    sw >= 9 ? shareware :
-    (cm >= 20 && hacx) ? (*gmission = pack_hacx, commercial) :
-    indetermined;
-
-  if (*gmode == retail && chex == 2)
-    *gmission = pack_chex;
-  if (*gmode == retail && rekkr)
-    *gmission = pack_rekkr;
-}
+    FILE *fp = fopen(iwadname, "rb");
 
-// jff 4/19/98 Add routine to check a pathname for existence as
-// a file or directory. If neither append .wad and check if it
-// exists as a file then. Else return non-existent.
+    if (fread(&header, 1, sizeof header, fp) == sizeof header)
+    {
+        fseek(fp, LONG(header.infotableofs), SEEK_SET);
 
-boolean WadFileStatus(char *filename,boolean *isdir)
-{
-  struct stat sbuf;
-  int i;
+        header.numlumps = LONG(header.numlumps);
 
-  *isdir = false;                 //default is directory to false
-  if (!filename || !*filename)    //if path NULL or empty, doesn't exist
-    return false;
+        while (header.numlumps)
+        {
+            if (fread(&lump, sizeof lump, 1, fp) == 1 &&
+                strncmp(lump.name, "MAP", 3) == 0)
+            {
+                count++;
+            }
 
-  if (!stat(filename,&sbuf))      //check for existence
-    {
-      *isdir=S_ISDIR(sbuf.st_mode); //if it does, set whether a dir or not
-      return true;                  //return does exist
+            header.numlumps--;
+        }
     }
 
-  i = strlen(filename);           //get length of path
-  if (i>=4)
-    if(!strnicmp(filename+i-4,".wad",4))
-      return false;               //if already ends in .wad, not found
+    fclose(fp);
 
-  strcat(filename,".wad");        //try it with .wad added
-  if (!stat(filename,&sbuf))      //if it exists then
+    return count;
+}
+
+static void CheckIWAD(const char *iwadname,
+                      GameMode_t *gmode,
+                      GameMission_t *gmission,  // joel 10/17/98 Final DOOM fix
+                      boolean *hassec)
+{
+    int i;
+    const char *name = M_BaseName(iwadname);
+
+    for (i = 0; i < arrlen(standard_iwads); ++i)
     {
-      if (S_ISDIR(sbuf.st_mode))  //but is a dir, then say we didn't find it
-        return false;
-      return true;                //otherwise return file found, w/ .wad added
+        if (!strcasecmp(name, standard_iwads[i].name))
+        {
+            *gmode = standard_iwads[i].mode;
+            *gmission = standard_iwads[i].mission;
+
+            if (*gmode == commercial)
+            {
+                *hassec = CountMaps(iwadname) < 31 ? false : true;
```

Comment:
I think it's enough to determinate IWAD by name. Do we need check for wolf levels here?

Reply:
We need this to set the `haswolflevels` variable.

Reply:
Yes, but we can remove `haswolflevels`, and fix it in Choco/Crispy way.

Reply:
We could check for the existence of the two map lumps once the IWAD is loaded, if that's what you mean.

Reply:
Yes, please.